### PR TITLE
r/sagemaker_image - new resource

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_sagemaker_image - new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerImage_'
--- PASS: TestAccAWSSagemakerImage_disappears (49.65s)
--- PASS: TestAccAWSSagemakerImage_basic (51.46s)
--- PASS: TestAccAWSSagemakerImage_displayName (109.36s)
--- PASS: TestAccAWSSagemakerImage_description (109.65s)
--- PASS: TestAccAWSSagemakerImage_tags (115.33s)
```

### CODE REVIEW

File: aws/resource_aws_sagemaker_image.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	needsUpdate := false
+
+	input := &sagemaker.UpdateImageInput{
+		ImageName: aws.String(d.Id()),
+	}
+
+	var deleteProperties []*string
+
+	if d.HasChange("role_arn") {
+		input.Description = aws.String(d.Get("role_arn").(string))
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:144]:
```suggestion
		input.RoleArn= aws.String(d.Get("role_arn").(string))
```

File: aws/resource_aws_sagemaker_image.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	log.Printf("[DEBUG] sagemaker Image create config: %#v", *input)
+	_, err := conn.CreateImage(input)
+	if err != nil {
+		return fmt.Errorf("error creating SageMaker Image: %w", err)
+	}
+
+	d.SetId(name)
+
+	if _, err := waiter.ImageCreated(conn, d.Id()); err != nil {
+		return fmt.Errorf("error waiting for sagemaker image (%s) to create: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:92]:
```suggestion
		return fmt.Errorf("error waiting for SageMaker Image (%s) to create: %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_image.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	d.Set("image_name", image.ImageName)
+	d.Set("arn", arn)
+	d.Set("role_arn", image.RoleArn)
+	d.Set("display_name", image.DisplayName)
+	d.Set("description", image.Description)
+
+	tags, err := keyvaluetags.SagemakerListTags(conn, arn)
+
+	if err != nil {
+		return fmt.Errorf("error listing tags for Sagemaker Image (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:123]:
```suggestion
		return fmt.Errorf("error listing tags for SageMaker Image (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_image.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+	if needsUpdate {
+		log.Printf("[DEBUG] sagemaker Image update config: %#v", *input)
+		_, err := conn.UpdateImage(input)
+		if err != nil {
+			return fmt.Errorf("error updating SageMaker Image: %w", err)
+		}
+
+		if _, err := waiter.ImageCreated(conn, d.Id()); err != nil {
+			return fmt.Errorf("error waiting for sagemaker image (%s) to update: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:175]:
```suggestion
			return fmt.Errorf("error waiting for SageMaker Image (%s) to update: %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_image.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+		}
+	}
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+
+		if err := keyvaluetags.SagemakerUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating Sagemaker Image (%s) tags: %s", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:183]:
```suggestion
			return fmt.Errorf("error updating SageMaker Image (%s) tags: %s", d.Id(), err)
```

File: aws/resource_aws_sagemaker_image.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+			return nil
+		}
+		return fmt.Errorf("error deleting SageMaker Image (%s): %w", d.Id(), err)
+	}
+
+	if _, err := waiter.ImageDeleted(conn, d.Id()); err != nil {
+		if isAWSErr(err, sagemaker.ErrCodeResourceNotFound, "No Image with the name") {
+			return nil
+		}
+		return fmt.Errorf("error waiting for sagemaker image (%s) to delete: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:208]:
```suggestion
		return fmt.Errorf("error waiting for SageMaker Image (%s) to delete: %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_image.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+	input := &sagemaker.UpdateImageInput{
+		ImageName: aws.String(d.Id()),
+	}
+
+	var deleteProperties []*string
+
+	if d.HasChange("role_arn") {
+		input.RoleArn= aws.String(d.Get("role_arn").(string))
+	}
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_image.go:145]: Fix formatting.

```suggestion
	if d.HasChange("role_arn") {
		input.RoleArn = aws.String(d.Get("role_arn").(string))
	}
```

Comment (DrFaust92) [reply aws/resource_aws_sagemaker_image.go]: ill add a test for role change, missed that.

**Comment by ewbankkit** [comment]
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerImage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerImage_ -timeout 120m
=== RUN   TestAccAWSSagemakerImage_basic
=== PAUSE TestAccAWSSagemakerImage_basic
=== RUN   TestAccAWSSagemakerImage_description
=== PAUSE TestAccAWSSagemakerImage_description
=== RUN   TestAccAWSSagemakerImage_displayName
=== PAUSE TestAccAWSSagemakerImage_displayName
=== RUN   TestAccAWSSagemakerImage_tags
=== PAUSE TestAccAWSSagemakerImage_tags
=== RUN   TestAccAWSSagemakerImage_disappears
=== PAUSE TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_basic
=== CONT  TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_tags
=== CONT  TestAccAWSSagemakerImage_displayName
=== CONT  TestAccAWSSagemakerImage_description
    resource_aws_sagemaker_image_test.go:94: Step 1/4 error: Error running apply: 2020/11/10 13:03:45 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for SageMaker Image (tf-acc-test-6701078467973104332) to create: unexpected state 'CREATE_FAILED', wanted target 'CREATED'. last error: %!s(<nil>)
        
        
=== CONT  TestAccAWSSagemakerImage_basic
    resource_aws_sagemaker_image_test.go:65: Step 1/2 error: Error running apply: 2020/11/10 13:03:46 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for SageMaker Image (tf-acc-test-633219264484407602) to create: unexpected state 'CREATE_FAILED', wanted target 'CREATED'. last error: %!s(<nil>)
        
        
=== CONT  TestAccAWSSagemakerImage_tags
    resource_aws_sagemaker_image_test.go:174: Step 1/4 error: Error running apply: 2020/11/10 13:03:46 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for SageMaker Image (tf-acc-test-1768421131957399270) to create: unexpected state 'CREATE_FAILED', wanted target 'CREATED'. last error: %!s(<nil>)
        
        
=== CONT  TestAccAWSSagemakerImage_displayName
    resource_aws_sagemaker_image_test.go:134: Step 1/4 error: Error running apply: 2020/11/10 13:03:45 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for SageMaker Image (tf-acc-test-3019142710389142362) to create: unexpected state 'CREATE_FAILED', wanted target 'CREATED'. last error: %!s(<nil>)
        
        
=== CONT  TestAccAWSSagemakerImage_disappears
    resource_aws_sagemaker_image_test.go:218: Step 1/1 error: Error running apply: 2020/11/10 13:03:46 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for SageMaker Image (tf-acc-test-8203763318322771741) to create: unexpected state 'CREATE_FAILED', wanted target 'CREATED'. last error: %!s(<nil>)
        
        
--- FAIL: TestAccAWSSagemakerImage_basic (12.83s)
--- FAIL: TestAccAWSSagemakerImage_disappears (12.83s)
--- FAIL: TestAccAWSSagemakerImage_description (13.25s)
--- FAIL: TestAccAWSSagemakerImage_tags (13.33s)
--- FAIL: TestAccAWSSagemakerImage_displayName (13.40s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	13.461s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

I assume some pre-requisite is missing.

**Comment by DrFaust92** [comment]
Didn't get this before. now getting this constantly. I tried to retry but same result. will take a look again

**Comment by DrFaust92** [comment]
So this is not as much as eventual consistency but having the right policy, ill leave a retry logic in place in any case and update docs on what is needed. AWS docs are not super clear but if you create a role via the Sagemaker console it adds a custom policy to read from some sagemaker buckets

**Comment by DrFaust92** [comment]
Ok, so this is even more annoying, sleeping before create solve it. simply retrying create (even for 10 mins) returns the same error (like its caching it on the backend?) and role can technically be passed as an update arguement but it never really updates it.

tests from latest changes
```
--- PASS: TestAccAWSSagemakerImage_disappears (98.66s)
--- PASS: TestAccAWSSagemakerImage_basic (117.88s)
--- PASS: TestAccAWSSagemakerImage_tags (164.56s)
--- PASS: TestAccAWSSagemakerImage_displayName (171.01s)
--- PASS: TestAccAWSSagemakerImage_description (173.37s)
```

**Comment by DrFaust92** [comment]
```
--- PASS: TestAccAWSSagemakerImage_disappears (100.42s)
--- PASS: TestAccAWSSagemakerImage_basic (118.25s)
--- PASS: TestAccAWSSagemakerImage_description (168.05s)
--- PASS: TestAccAWSSagemakerImage_displayName (168.89s)
--- PASS: TestAccAWSSagemakerImage_tags (172.61s)
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerImage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerImage_ -timeout 120m
=== RUN   TestAccAWSSagemakerImage_basic
=== PAUSE TestAccAWSSagemakerImage_basic
=== RUN   TestAccAWSSagemakerImage_description
=== PAUSE TestAccAWSSagemakerImage_description
=== RUN   TestAccAWSSagemakerImage_displayName
=== PAUSE TestAccAWSSagemakerImage_displayName
=== RUN   TestAccAWSSagemakerImage_tags
=== PAUSE TestAccAWSSagemakerImage_tags
=== RUN   TestAccAWSSagemakerImage_disappears
=== PAUSE TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_basic
=== CONT  TestAccAWSSagemakerImage_tags
=== CONT  TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_displayName
=== CONT  TestAccAWSSagemakerImage_description
--- PASS: TestAccAWSSagemakerImage_disappears (80.67s)
--- PASS: TestAccAWSSagemakerImage_basic (81.45s)
--- PASS: TestAccAWSSagemakerImage_description (103.85s)
--- PASS: TestAccAWSSagemakerImage_tags (108.45s)
--- PASS: TestAccAWSSagemakerImage_displayName (110.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	110.319s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerImage_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerImage_ -timeout 120m
=== RUN   TestAccAWSSagemakerImage_basic
=== PAUSE TestAccAWSSagemakerImage_basic
=== RUN   TestAccAWSSagemakerImage_description
=== PAUSE TestAccAWSSagemakerImage_description
=== RUN   TestAccAWSSagemakerImage_displayName
=== PAUSE TestAccAWSSagemakerImage_displayName
=== RUN   TestAccAWSSagemakerImage_tags
=== PAUSE TestAccAWSSagemakerImage_tags
=== RUN   TestAccAWSSagemakerImage_disappears
=== PAUSE TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_basic
=== CONT  TestAccAWSSagemakerImage_tags
=== CONT  TestAccAWSSagemakerImage_disappears
=== CONT  TestAccAWSSagemakerImage_displayName
=== CONT  TestAccAWSSagemakerImage_description
--- PASS: TestAccAWSSagemakerImage_disappears (76.16s)
--- PASS: TestAccAWSSagemakerImage_basic (78.60s)
--- PASS: TestAccAWSSagemakerImage_description (98.61s)
--- PASS: TestAccAWSSagemakerImage_tags (99.68s)
--- PASS: TestAccAWSSagemakerImage_displayName (101.98s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	103.869s
```

Note that this resource is part of SageMaker Studio which is not available in GovCloud. This will require work to skip these tests when run in that region. This is tracked here: #16968

**Pull request merged**